custom_entity/Field/FieldWidget/DefaultEntityDurationWidget.php

```diff
@@ -0,0 +1,214 @@
+      '#default_value' =>
+        $items[$delta]->end_date ?
+          date('Y-m-d', $items[$delta]->end_date) : '',
+      '#prefix' => '<span class="error-message"></span>',
+      '#element_validate' => [
+        [$this, 'validateDuration'],
+      ]
+    ];
+
+
+    if ($type != 'textfield')
+      return $element;
+
+    /** If type is textfield add ajax validation for elements */
+    $element['start_date']['#ajax'] = [
+      'callback' => [$this, 'validateAjax'],
+      'event' => 'change',
+    ];
+    array_unshift($element['start_date']['#element_validate'],
+      [$this, 'validateDateFormat']);
+
+    $element['end_date']['#ajax'] = [
+      'callback' => [$this, 'validateAjax'],
+      'event' => 'change',
+    ];
+    array_unshift($element['end_date']['#element_validate'],
+      [$this, 'validateDateFormat']);
+
+    return $element;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function defaultSettings() {
+    return ['date_format' => 'date'] + parent::defaultSettings();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function settingsForm(array $form, FormStateInterface $form_state) {
+    $form = parent::settingsForm($form, $form_state);
+    $form['date_format'] = [
+      '#title' => $this->t('Format of the date'),
+      '#type' => 'select',
+      '#options' => [
+        'date' => $this->t('Calendar'),
+        'textfield' => $this->t('Text')
+      ],
+      '#default_value' => $this->getSetting('date_format'),
+    ];
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function massageFormValues(array $values,
+    array $form,
+    FormStateInterface $form_state
+  ) {
+    $values = parent::massageFormValues($values, $form, $form_state);
+    foreach ($values as &$item) {
+      if (isset($item['start_date'])) {
+        $item['start_date'] = strtotime($item['start_date']);
+      }
+
+      if (isset($item['end_date'])) {
+        $item['end_date'] = strtotime($item['end_date']);
+      };
+    }
+    return $values;
+  }
+
+  /**
+   * Validate date format by Ajax.
+   *
+   * @param array $element
+   * @return AjaxResponse
+   */
+  public function validateAjax(array &$form, FormStateInterface $form_state)
+  {
+    $response = new AjaxResponse();
+    $element = $form_state->getTriggeringElement();
+    if(!strtotime($element['#value'])) {
+      $response->addCommand(new HtmlCommand(
+        '.error-message',
+        $this->t(self::INVALID_DATE_FORMAT_ERROR_MESSAGE)));
+    }
+
+    return $response;
+  }
+
+  /**
+   * Validate date format and duration
+   *
+   * @param $element
+   * @param FormStateInterface $formState
+   * @param $form
+   */
+  public function validateDuration(array $element, FormStateInterface $formState, $form)
+  {
+    $dates = $this->getElementsByNames(['start_date', 'end_date'], $form);
+    if (!$this->isValidDuration(strtotime($dates['start_date']['#value']), strtotime($dates['end_date']['#value'])))
```

Comment:
Can't we just change this structure to
strtotime($dates['start_date']['#value']) < strtotime($dates['end_date']['#value'])